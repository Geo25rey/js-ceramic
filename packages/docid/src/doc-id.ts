import CID from 'cids';
import multibase from 'multibase';
import doctypes from './doctype-table';
import varint from 'varint';
import uint8ArrayConcat from 'uint8arrays/concat';
import uint8ArrayToString from 'uint8arrays/to-string';
import { DEFAULT_BASE, DOCID_CODEC } from './constants';

const getKey = (obj: { [key: string]: number }, value: number): string | undefined => {
  for (const [k, v] of Object.entries(obj)) {
    if (v === value) return k;
  }
};

// Definition
// '<multibase-prefix><multicodec-docid><doctype><genesis-cid-bytes>'
// '<multibase-prefix><multicodec-docid><doctype><genesis-cid-bytes><commit-cid-bytes>'

export class DocID {
  /**
   * Create a new DocID.
   *
   * @param {string|number}      doctype
   * @param {CID|string}         cid
   * @param {CID|string}         commit CID. Pass '0' as shorthand for the genesis commit.
   * @param {string}             [multibaseName = 'base36']
   *
   * @example
   * new DocID(<docType>, <CID>|<cidStr>)
   * new DocID(<docType>, <CID>|<cidStr>, <CommitCID>|<CommitCidStr>, <multibaseName>)
   */

  readonly #doctype: number;
  readonly #cid: CID;
  private _bytes: Uint8Array;

  constructor(doctype: string | number, cid: CID | string) {
    if (!cid) throw new Error('constructor: cid required');
    this.#doctype = typeof doctype === 'string' ? doctypes[doctype] : doctype;
    if (!this.#doctype && this.#doctype !== 0) throw new Error('constructor: doctype required');
    this.#cid = typeof cid === 'string' ? new CID(cid) : cid;
  }

  /**
   * Copies the given DocID and returns a copy of it, optionally changing the commit to the one provided
   * @param other
   */
  static fromOther(other: DocID): DocID {
    return new DocID(other.type, other.cid);
  }

  static fromBytes(bytes: Uint8Array): DocID {
    const docCodec = varint.decode(bytes);
    if (docCodec !== DOCID_CODEC) throw new Error('fromBytes: invalid docid, does not include docid codec');
    bytes = bytes.slice(varint.decode.bytes);
    const docType = varint.decode(bytes);
    bytes = bytes.slice(varint.decode.bytes);

    let cid;

    try {
      cid = new CID(bytes);
    } catch (e) {
      // Includes commit
      const cidLength = DocID._genesisCIDLength(bytes);
      cid = new CID(bytes.slice(0, cidLength));
    }

    return new DocID(docType, cid);
  }

  static _genesisCIDLength(bytes: Uint8Array): number {
    let offset = 0;

    varint.decode(bytes); // cid version
    offset += varint.decode.bytes;

    varint.decode(bytes.slice(offset)); // cid codec
    offset += varint.decode.bytes;

    varint.decode(bytes.slice(offset)); //mh codec
    offset += varint.decode.bytes;

    const length = varint.decode(bytes.slice(offset)); //mh length
    return offset + length + 1;
  }

  static fromString(docId: string): DocID {
    docId = docId.split('ceramic://').pop();
    // Likely temp, remove legacy once all ceramic update, but should make updating easier
    docId = docId.split('/ceramic/').pop();
    if (docId.includes('commit')) {
      docId = docId.split('?')[0];
    }
    const bytes = multibase.decode(docId);
    return DocID.fromBytes(bytes);
  }

  /**
   * Get doc type code
   *
   * @returns {number}
   * @readonly
   */
  get type(): number {
    return this.#doctype;
  }

  /**
   * Get doc type name
   *
   * @returns {string}
   * @readonly
   */
  get typeName(): string {
    const name = getKey(doctypes, this.#doctype);
    if (!name) throw new Error('docTypeName: no registered name available');
    return name;
  }

  /**
   * Get CID
   *
   * @returns {CID}
   * @readonly
   */
  get cid(): CID {
    return this.#cid;
  }

  /**
   * Get CID codec name string
   *
   * @returns {string}
   * @readonly
   */
  get codec(): string {
    return this.#cid.codec;
  }

  /**
   * Get bytes of DocId
   *
   * @returns {Uint8Array}
   * @readonly
   */
  get bytes(): Uint8Array {
    if (this._bytes == null) {
      const codec = varint.encode(DOCID_CODEC);
      const doctype = varint.encode(this.type);

      this._bytes = uint8ArrayConcat([codec, doctype, this.cid.bytes]);
    }
    return this._bytes;
  }

  /**
   * Compare equality with another DocID.
   *
   * @param   {DocID}   other
   * @returns {bool}
   */
  equals(other: DocID | Uint8Array | string): boolean {
    let otherDocID;
    if (typeof other === 'string') otherDocID = DocID.fromString(other);
    else if (other instanceof Uint8Array) {
      otherDocID = DocID.fromBytes(other);
    } else {
      otherDocID = other;
    }

    return this.type === otherDocID.type && this.cid.equals(otherDocID.cid);
  }

  /**
   * Encode the DocID into a string.
   */
  toString(): string {
    return uint8ArrayToString(multibase.encode(DEFAULT_BASE, this.bytes));
  }

  /**
   * Encode the DocID into a base36 url
   *
   * @returns {string}
   */
  toUrl(): string {
    return `ceramic://${this.toString()}`;
  }

  /**
   * DocId(k3y52l7mkcvtg023bt9txegccxe1bah8os3naw5asin3baf3l3t54atn0cuy98yws)
   *
   * @returns {String}
   */
  [Symbol.for('nodejs.util.inspect.custom')](): string {
    return `DocID(${this.toString()})`;
  }

  [Symbol.toPrimitive](): string | Uint8Array {
    return this.toString();
  }

  /**
   * Determine if given DocID, DocID string or bytes is a valid DocID
   *
   * @param   {any}     other
   * @returns {Boolean}
   */
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  static isDocID(other: any): boolean {
    try {
      if (typeof other === 'string') {
        DocID.fromString(other);
        return true;
      } else if (other instanceof Uint8Array) {
        DocID.fromBytes(other);
        return true;
      } else {
        return (other.type || other.type === 0) && Boolean(other.cid);
      }
    } catch (e) {
      return false;
    }
  }
}
