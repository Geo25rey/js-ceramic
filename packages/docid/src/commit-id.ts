import CID from 'cids';
import multibase from 'multibase';
import * as doctypes from './doctypes';
import varint from 'varint';
import uint8ArrayConcat from 'uint8arrays/concat';
import uint8ArrayToString from 'uint8arrays/to-string';
import { DEFAULT_BASE, DOCID_CODEC } from './constants';
import { readCid, readVarint } from './reading-bytes';
import { DocID } from './doc-id';

// Definition
// '<multibase-prefix><multicodec-docid><doctype><genesis-cid-bytes>'
// '<multibase-prefix><multicodec-docid><doctype><genesis-cid-bytes><commit-cid-bytes>'

function fromOther(other: CommitId): CommitId {
  return new CommitId(other.type, other.cid, other.commit);
}

function fromBytes(bytes: Uint8Array): CommitId {
  const [docCodec, docCodecRemainder] = readVarint(bytes);
  if (docCodec !== DOCID_CODEC) throw new Error('fromBytes: invalid docid, does not include docid codec');
  const [doctype, doctypeRemainder] = readVarint(docCodecRemainder);
  const [base, baseRemainder] = readCid(doctypeRemainder);
  if (baseRemainder.length === 0) {
    // No commit AKA DocID
    return new CommitId(doctype, base);
  } else if (baseRemainder.length === 1) {
    // Zero commit
    return new CommitId(doctype, base, baseRemainder[0]);
  } else {
    // Commit
    const [commit] = readCid(baseRemainder);
    return new CommitId(doctype, base, commit);
  }
}

function parseCID(input: any) {
  try {
    return new CID(input);
  } catch {
    return undefined;
  }
}

function parseCommit(base: CID, commit: CID | string | number = null): CID | null {
  if (!commit) return null;

  const commitCID = parseCID(commit);
  if (commitCID) {
    // CID-like
    if (base.equals(commitCID)) {
      return null;
    } else {
      return commitCID;
    }
  } else if (String(commit) === '0') {
    // Zero as number or string
    return null;
  } else {
    throw new Error('Cannot specify commit as a number except to request commit 0 (the genesis commit)');
  }
}

function fromString(input: string): CommitId {
  const protocolFree = input.replace('ceramic://', '').replace('/ceramic/', '');
  if (protocolFree.includes('commit')) {
    const commit = protocolFree.split('?')[1].split('=')[1];
    const base = protocolFree.split('?')[0];
    return fromBytes(multibase.decode(base)).travel(commit);
  } else {
    return fromBytes(multibase.decode(protocolFree));
  }
}

export class CommitId {
  readonly #doctype: number;
  readonly #cid: CID;
  readonly #commit: CID | null;
  private _bytes: Uint8Array;

  /**
   * Copies the given DocID and returns a copy of it, optionally changing the commit to the one provided
   * @param other
   * @param commit
   */
  static fromOther = fromOther;
  static fromBytes = fromBytes;
  static fromString = fromString;

  /**
   * Create a new DocID.
   *
   * @param {string|number}      doctype
   * @param {CID|string}         cid
   * @param {CID|string}         commit CID. Pass '0' as shorthand for the genesis commit.
   *
   * @example
   * new DocID(<docType>, <CID>|<cidStr>)
   * new DocID(<docType>, <CID>|<cidStr>, <CommitCID>|<CommitCidStr>, <multibaseName>)
   */
  constructor(doctype: string | number, cid: CID | string, commit: CID | string | number = null) {
    if (!doctype && doctype !== 0) throw new Error('constructor: doctype required');
    if (!cid) throw new Error('constructor: cid required');
    this.#doctype = typeof doctype === 'string' ? doctypes.indexByName(doctype) : doctype;
    this.#cid = typeof cid === 'string' ? new CID(cid) : cid;
    this.#commit = parseCommit(this.#cid, commit);
  }

  /**
   * Get base docID, always returns without commit
   *
   * @returns {DocID}
   * @readonly
   */
  get baseID(): DocID {
    return new DocID(this.#doctype, this.#cid);
  }

  /**
   * Get doc type code
   *
   * @returns {number}
   * @readonly
   */
  get type(): number {
    return this.#doctype;
  }

  /**
   * Get doc type name
   *
   * @returns {string}
   * @readonly
   */
  get typeName(): string {
    return doctypes.nameByIndex(this.#doctype);
  }

  /**
   * Get CID
   *
   * @returns {CID}
   * @readonly
   */
  get cid(): CID {
    return this.#cid;
  }

  /**
   * Get Commit CID
   *
   * @returns {CID}
   * @readonly
   */
  get commit(): CID {
    return this.#commit || this.#cid;
  }

  /**
   * Get CID codec name string
   *
   * @returns {string}
   * @readonly
   */
  get codec(): string {
    return this.#cid.codec;
  }

  /**
   * Get bytes of DocId
   *
   * @returns {Uint8Array}
   * @readonly
   */
  get bytes(): Uint8Array {
    if (this._bytes == null) {
      const codec = varint.encode(DOCID_CODEC);
      const doctype = varint.encode(this.type);

      const commitBytes = this.#commit?.bytes || new Uint8Array(0);
      this._bytes = uint8ArrayConcat([codec, doctype, this.cid.bytes, commitBytes]);
    }
    return this._bytes;
  }

  travel(commit: CID | string | number): CommitId {
    return new CommitId(this.#doctype, this.#cid, commit);
  }

  /**
   * Compare equality with another DocID.
   *
   * @param   {DocID}   other
   */
  equals(other: CommitId): boolean {
    return this.type === other.type && this.cid.equals(other.cid) && this.commit.equals(other.commit);
  }

  /**
   * Encode the DocID into a string.
   */
  toString(): string {
    return uint8ArrayToString(multibase.encode(DEFAULT_BASE, this.bytes));
  }

  /**
   * Encode the DocID into a base36 url
   *
   * @returns {string}
   */
  toUrl(): string {
    return `ceramic://${this.toString()}`;
  }

  /**
   * If genesis commit:
   * CommitID(k3y52l7mkcvtg023bt9txegccxe1bah8os3naw5asin3baf3l3t54atn0cuy98yws, 0)
   *
   * If commit:
   * CommitID(k3y52l7mkcvtg023bt9txegccxe1bah8os3naw5asin3baf3l3t54atn0cuy98yws, bagjqcgzaday6dzalvmy5ady2m5a5legq5zrbsnlxfc2bfxej532ds7htpova)
   *
   * @returns {String}
   */
  [Symbol.for('nodejs.util.inspect.custom')](): string {
    if (this.#commit) {
      return `CommitID(${this.#cid.toString()}, ${this.#commit.toString()})`;
    } else {
      return `CommitID(${this.#cid.toString()}, 0)`;
    }
  }

  [Symbol.toPrimitive](): string | Uint8Array {
    return this.toString();
  }
}
