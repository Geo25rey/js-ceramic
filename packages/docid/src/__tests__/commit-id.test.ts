import * as util from 'util';
import { CommitId } from '../commit-id';
import CID from 'cids';
import * as multibase from 'multibase';
import { DocID } from '../doc-id';

const BASE_CID_STRING = 'bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a';
const BASE_CID = new CID(BASE_CID_STRING);
const COMMIT_CID_STRING = 'bagjqcgzaday6dzalvmy5ady2m5a5legq5zrbsnlxfc2bfxej532ds7htpova';
const COMMIT_CID = new CID(COMMIT_CID_STRING);

const DOC_ID_STRING = 'kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s';
const DOC_ID_BYTES = multibase.decode(DOC_ID_STRING);
const DOC_ID_URL = 'ceramic://kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s';
const DOC_ID_URL_LEGACY = '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s';

const DOC_ID_WITH_COMMIT =
  'k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju';
const DOC_ID_WITH_COMMIT_BYTES = multibase.decode(DOC_ID_WITH_COMMIT);

const DOC_ID_WITH_0_COMMIT = 'k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4';
const DOC_ID_WITH_0_COMMIT_BYTES = multibase.decode(DOC_ID_WITH_0_COMMIT);

const DOC_ID_WITH_COMMIT_LEGACY =
  '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s?commit=bagjqcgzaday6dzalvmy5ady2m5a5legq5zrbsnlxfc2bfxej532ds7htpova';
const DOC_ID_WITH_0_COMMIT_LEGACY = '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s?commit=0';

describe('constructor', () => {
  test('create by parts (type:int, cid:cid)', () => {
    const type = 0;
    const docid = new CommitId(type, BASE_CID);

    expect(docid.type).toEqual(type);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('create by parts (type:string, cid:string, commit:null)', () => {
    const type = 'tile';
    const docid = new CommitId(type, BASE_CID_STRING, null);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('create by parts (type:string, cid:string, commit:string)', () => {
    const type = 'tile';
    const docid = new CommitId(type, BASE_CID_STRING, BASE_CID_STRING);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('unknown doctype name', () => {
    expect(() => new CommitId('invalid-garbage', BASE_CID_STRING, BASE_CID_STRING)).toThrow();
  });
});

describe('#travel', () => {
  const commitId = new CommitId('tile', BASE_CID, COMMIT_CID);

  test('to number 0', () => {
    const traveller = commitId.travel(0);
    expect(traveller.commit).toEqual(BASE_CID);
  });
  test('to number 1', () => {
    expect(() => commitId.travel(1)).toThrowErrorMatchingSnapshot();
  });
  test('to commit CID', () => {
    const commitId = new CommitId('tile', BASE_CID);
    const traveller = commitId.travel(COMMIT_CID);
    expect(traveller.commit).toEqual(COMMIT_CID);
  });
  test('to commit CID as string', () => {
    const commitId = new CommitId('tile', BASE_CID);
    const traveller = commitId.travel(COMMIT_CID_STRING);
    expect(traveller.commit).toEqual(COMMIT_CID);
  });
  test('to garbage string', () => {
    expect(() => commitId.travel('garbage')).toThrow();
  });
});

describe('.fromBytes', () => {
  test('create from bytes', () => {
    const docid = CommitId.fromBytes(DOC_ID_BYTES);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('invalid doc id', () => {
    expect(() => CommitId.fromBytes(BASE_CID.bytes)).toThrow();
  });

  test('create from bytes inlcuding commit', () => {
    const docid = CommitId.fromBytes(DOC_ID_WITH_COMMIT_BYTES);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(COMMIT_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('create from bytes inlcuding commit 0', () => {
    const docid = CommitId.fromBytes(DOC_ID_WITH_0_COMMIT_BYTES);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('roundtrip docID bytes', () => {
    const docid = new CommitId('tile', BASE_CID_STRING);
    const docid2 = CommitId.fromBytes(docid.bytes);
    expect(docid.toString()).toEqual(docid2.toString());
    expect(docid.bytes).toEqual(docid2.bytes);
  });

  test('roundtrip docID bytes with commit', () => {
    const docid = new CommitId('tile', BASE_CID_STRING, COMMIT_CID_STRING);
    const docid2 = CommitId.fromBytes(docid.bytes);
    expect(docid.toString()).toEqual(docid2.toString());
    expect(docid.bytes).toEqual(docid2.bytes);
  });

  test('roundtrip docID bytes with commit', () => {
    const docid = new CommitId('tile', BASE_CID_STRING, 0);
    const docid2 = CommitId.fromBytes(docid.bytes);
    expect(docid.toString()).toEqual(docid2.toString());
    expect(docid.bytes).toEqual(docid2.bytes);
  });
});

describe('.fromString', () => {
  test('create from string', () => {
    const docid = CommitId.fromString(DOC_ID_STRING);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('create from string including commit', () => {
    const docid = CommitId.fromString(DOC_ID_WITH_COMMIT);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(COMMIT_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('create from string including commit 0', () => {
    const docid = CommitId.fromString(DOC_ID_WITH_0_COMMIT);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('create from url string', () => {
    const docid = CommitId.fromString(DOC_ID_URL);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('create from legacy string "/ceramic/"', () => {
    const docid = CommitId.fromString(DOC_ID_URL_LEGACY);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('create from legacy string "/ceramic/" with commit param "?commit="', () => {
    const docid = CommitId.fromString(DOC_ID_WITH_COMMIT_LEGACY);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(COMMIT_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('create from legacy string "/ceramic/" with commit param "?commit=0"', () => {
    const docid = CommitId.fromString(DOC_ID_WITH_0_COMMIT_LEGACY);

    expect(docid.type).toEqual(0);
    expect(docid.cid).toEqual(BASE_CID);
    expect(docid.commit).toEqual(BASE_CID);
    expect(docid.toString()).toMatchSnapshot();
  });

  test('roundtrip docID string', () => {
    const docid = new CommitId('tile', BASE_CID_STRING);
    const docid2 = CommitId.fromString(docid.toString());
    expect(docid.toString()).toEqual(docid2.toString());
  });

  test('roundtrip docID string with commit', () => {
    const docid = new CommitId('tile', BASE_CID_STRING, COMMIT_CID_STRING);
    const docid2 = CommitId.fromString(docid.toString());
    expect(docid.toString()).toEqual(docid2.toString());
  });

  test('roundtrip docID string with commit 0', () => {
    const docid = new CommitId('tile', BASE_CID_STRING, 0);
    const docid2 = CommitId.fromString(docid.toString());
    expect(docid.toString()).toEqual(docid2.toString());
  });
});

test('.bytes', () => {
  const docid = new CommitId('tile', BASE_CID_STRING);
  const bytes = docid.bytes;
  expect(bytes).toBeDefined();
  expect(bytes instanceof Uint8Array).toEqual(true);
  expect(bytes).toMatchSnapshot();
});

test('.typeName if registered', () => {
  const docid = new CommitId('tile', BASE_CID_STRING, null);
  expect(docid.typeName).toEqual('tile');
  const docid2 = new CommitId(10, BASE_CID_STRING);
  expect(() => docid2.typeName).toThrowErrorMatchingSnapshot();
});

test('.toString()', () => {
  const docid = new CommitId('tile', BASE_CID_STRING);
  const str = docid.toString();
  expect(str).toBeDefined();
  expect(str).toMatchSnapshot();
});

test('.toUrl()', () => {
  const docid = new CommitId('tile', BASE_CID_STRING);
  const str = docid.toUrl();
  expect(str).toBeDefined();
  expect(str.includes('ceramic://')).toEqual(true);
  expect(str).toMatchSnapshot();
});

describe('#equals', () => {
  test('equals other DocID', () => {
    const docid = new CommitId('tile', BASE_CID_STRING);
    const docid2 = new CommitId('tile', BASE_CID_STRING);
    const docid3 = new CommitId('caip10-link', BASE_CID_STRING);
    expect(docid.equals(docid2)).toEqual(true);
    expect(docid.equals(docid3)).toEqual(false);
  });
});

describe('nodejs inspect', () => {
  test('genesis commit', () => {
    const docid = new CommitId('tile', BASE_CID_STRING);
    expect(util.inspect(docid)).toMatchSnapshot();
  });
  test('commit', () => {
    const docid = new CommitId('tile', BASE_CID_STRING, COMMIT_CID_STRING);
    expect(util.inspect(docid)).toMatchSnapshot();
  });
});

test('to primitive', () => {
  const docid = new CommitId('tile', BASE_CID_STRING, COMMIT_CID_STRING);
  expect(`${docid}`).toEqual(docid.toString());
  expect(+docid).toBeNaN();
  expect(docid + '').toEqual(docid.toString());
});
