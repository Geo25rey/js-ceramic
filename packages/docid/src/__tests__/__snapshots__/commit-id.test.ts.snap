// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#travel to number 1 1`] = `"Cannot specify commit as a number except to request commit 0 (the genesis commit)"`;

exports[`.bytes 1`] = `
Uint8Array [
  206,
  1,
  0,
  1,
  133,
  1,
  18,
  32,
  84,
  179,
  109,
  86,
  78,
  173,
  242,
  220,
  63,
  166,
  43,
  94,
  52,
  240,
  80,
  102,
  174,
  138,
  255,
  216,
  30,
  185,
  101,
  36,
  150,
  57,
  240,
  167,
  85,
  209,
  179,
  188,
]
`;

exports[`.fromBytes create from bytes  1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromBytes create from bytes inlcuding commit 0 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromBytes create from bytes inlcuding commit 1`] = `"k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju"`;

exports[`.fromBytes create from bytes with commit (0) matching cid 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromBytes create from bytes with commit 1`] = `"k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju"`;

exports[`.fromString create from legacy string "/ceramic/" 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromString create from legacy string "/ceramic/" with commit param "?commit=" 1`] = `"k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju"`;

exports[`.fromString create from legacy string "/ceramic/" with commit param "?commit=0" 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromString create from string 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromString create from string including commit 0 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromString create from string including commit 1`] = `"k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju"`;

exports[`.fromString create from url string 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.toString() 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.toUrl() 1`] = `"ceramic://kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.typeName if registered 1`] = `"No doctype registered for index 10"`;

exports[`DocID create by parts (type:int, cid:cid) 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create by parts (type:string, cid:string, commit:null) 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create by parts (type:string, cid:string, commit:string) 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`constructor create by parts (type:int, cid:cid) 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`constructor create by parts (type:string, cid:string, commit:null) 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`constructor create by parts (type:string, cid:string, commit:string) 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`nodejs inspect commit 1`] = `"CommitID(bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a, bagjqcgzaday6dzalvmy5ady2m5a5legq5zrbsnlxfc2bfxej532ds7htpova)"`;

exports[`nodejs inspect genesis commit 1`] = `"CommitID(bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a, 0)"`;
